City-of-Helsinki/yjdh
# HL-553, HL-566 | Update mock flag related permission handling & fix moving application to handling

## Description :sparkles:

### Shared backend: Add is_staff & is_superuser support to OIDC mock user 

Add settings to control OIDC mock user's is_staff & is_superuser values:
 - settings.OIDC_MOCK_USER_IS_STAFF -> OIDC mock user's is_staff
   - If the setting is not set at all then will default to False
 - settings.OIDC_MOCK_USER_IS_SUPERUSER -> OIDC mock user's is_superuser
   - If the setting is not set at all then will default to False

The OIDC mock user is created by MockAuthenticationRequestView which is
mapped to oidc's authenticate endpoint
(See backend/shared/shared/oidc/urls.py) if NEXT_PUBLIC_MOCK_FLAG is set

Refs HL-566
 
### HL-Backend: Update NEXT_PUBLIC_MOCK_FLAG related permission handling 

BaseApplicationViewSet.get_queryset:
 - Refactor logical structure for readability, should not change effect

ApplicationStatus.is_editable_status:
 - Previously NEXT_PUBLIC_MOCK_FLAG allowed to edit all statuses. Now
   NEXT_PUBLIC_MOCK_FLAG allows to edit both handler's and applicant's
   statuses.
 - Update related test_is_editable_status_with_anonymous_user test

UserSerializer.is_terms_of_service_approval_needed:
 - Don't require terms of service to be approved when
   NEXT_PUBLIC_MOCK_FLAG is on

Refs HL-566
 
### Shared backend: Conditionally mock endpoints using MockEnabledProxyView 

Add MockEnabledProxyView and tests:
 - A view that changes on-the-fly between the real and mocked view based
   on the current value of NEXT_PUBLIC_MOCK_FLAG

Refs HL-566
 
### HL-Backend: Don't validate pay subsidies in RECEIVED to HANDLING change 

PaysubsidySerializer:
 - Make start_date and end_date validation errors specific to the fields
 - Skip start_date and end_date validation when application's status is
   changed from RECEIVED to HANDLING

Tests:
 - Add RECEIVED to HANDLING status change case to
   test_application_status_change_as_handler
 - Add test_ignore_pay_subsidy_dates_in_received_to_handling_transition
   test to check that pay subsidies' start and end dates can be empty
   when application's status changes from RECEIVED to HANDLING
 - Move duplicated test code to _set_two_pay_subsidies_with_empty_dates

Translations:
 - Update translations:
   - "python manage.py makemessages --no-location -l fi -l sv -l en"
 - Add Finnish and Swedish translations for
   - "Start date cannot be empty"
   - "End date cannot be empty"

Refs HL-553

## Issues :bug:

HL-553, HL-566

## Testing :alembic:

## Screenshots :camera_flash:

## Additional notes :spiral_notepad:
